### PR TITLE
Support disabling the sbt task linter

### LINKED ISSUE #3266: Reconsider failing `.value` inside lambda expression
## steps

```scala
    test in Test := {
      List(1, 2, 3) foreach { x =>
        streams.value.log.info(s"foo $x")
      }
      (test in Test).value
    }
```

## problem

```
build.sbt:17: error: The evaluation of `streams` inside an anonymous function is prohibited.

Problem: Task invocations inside anonymous functions are evaluated independently of whether the anonymous function is invoked or not.

Solution:
  1. Make `streams` evaluation explicit outside of the function body if you don't care about its evaluation.
  2. Use a dynamic task to evaluate `streams` and pass that value as a parameter to an anonymous function.

        streams.value.log.info(s"foo $x")
        ^
[error] sbt.compiler.EvalException: Type error in expression
[error] Use 'last' for the full log.
```

## expectation

The task does exactly what I want it to do. It should work without warnings or error.

## notes

In general, an error (or even a warning) should be displayed when we are 90% certain that the build is doing something against what the user has intended. In this case, I don't think it is.

Ref https://github.com/lucidsoftware/neo-sbt-scalafmt/pull/11, https://github.com/sbt/sbt/pull/3216/commits/ca3acc4e52668c77a1f2dfa0cc247a45936cf7bd

/cc @jvican

### PR DESCRIPTION
**Pull request by eatkins**
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines
This PR resolves #3266. It also does some minor clean up of the TaskLinterDSL to make it more maintainable in general.

### CODE REVIEW

**Comment by eatkins** [comment]
cc @dwijnand @eed3si9n. I whipped this up fairly quickly and didn't give a ton of thought to the names and project layout.

**Comment by dwijnand** [comment]
I reviewed it _really_ really quick and it looked good, but I also approve of fixing that issue in general.  Thanks, Ethan.

File: main-settings/src/main/scala/sbt/dsl/LinterLevel.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/*
+ * sbt
+ * Copyright 2011 - 2018, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.dsl
+
+sealed trait LinterLevel
```
Comment (eatkins) [inline main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:47]: I considered contraband for this, but wasn't sure if there is a way to put the (prioritized) implicit instances in the companion object.

Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:47]: You can use https://www.scala-sbt.org/contraband/schema.html#Extra+code

```
  #xcompanioninterface CompanionInterface1
```

**Comment by eed3si9n** [comment]
This is awesome. Thanks @eatkins!

File: main-settings/src/main/scala/sbt/dsl/LinterLevel.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+object LinterLevel extends LinterLevelLowPriority {
+  implicit case object Warn extends LinterLevel
```
Comment (eed3si9n) [inline main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:54]: There should be an explanation of why this is made implicit, and how this is meant to be used.

**Review by eed3si9n — CHANGES_REQUESTED** [review]

File: main-settings/src/main/scala/sbt/dsl/LinterLevel.scala (hunk 2 above)
Comment (eatkins) [reply main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:54]: I force pushed an update that added scaladoc to these.

**Review by eed3si9n — APPROVED** [review]

**Comment by dwijnand** [comment]
Where did that fix come from?!  Very nice.

If you want a more common source of dynamic tasks put some projects in a seq and then map the seq, scope a key and .value that.  Users try to do that all the time.

```scala
Seq(foo, bar, baz).map(p =>
  (bark in p).value
)
```

**Comment by eatkins** [comment]
Funnily enough, this is all related #3718. I'm working on trying to implement the taskSources idea in the comment at the bottom. This requires traversing the list of dependencies for each task that is being watched to extract all of the taskSources dependencies. While working on that, I realized that dynamic tasks are a huge problem because we lose the dependency information for dynamic tasks. I started working on modifying the Def.taskDyn macro to try and capture all of the dependencies for the Def.initialize instances that are referenced in the dynamic task definition. This forced me to get quite familiar with the TaskLinterDSL at which point I decided to go ahead and fix #3110 and #3266, both of which had been irritating me for a long time.

**Pull request merged**